File: LeetCode/543.Diameter_of_Binary_Tree/Mario/543.Diameter_of_Binary_Tree.go
var diameter int
func diameterOfBinaryTree(root *TreeNode) int {
    diameter = 0
    longestPath(root)
    return diameter
}

func longestPath(node *TreeNode) int{
    if node == nil {
        return 0
    }

    left := longestPath(node.Left)
    right := longestPath(node.Right)

    diameter = max(diameter, left + right)

    return max(left, right) + 1
}

File: LeetCode/543.Diameter_of_Binary_Tree/543.Diameter_of_Binary_Tree.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
class Solution {
    func diameterOfBinaryTree(_ root: TreeNode?) -> Int {
        var diameter = 0
        
        func helper(_ node: TreeNode?) -> Int {
            guard let node = node else {
                return 0
            }
            
            let left = helper(node.left)
            let right = helper(node.right)

            diameter = max(diameter, left + right)

            return max(left, right) + 1
        }

        helper(root)

        return diameter
    }
}